src/tools/chip-cert/Cmd_PrintCD.cpp

```diff
@@ -0,0 +1,277 @@
+    "Print a CHIP certification declaration (CD) content.\n"
+    "\n"
+    "ARGUMENTS\n"
+    "\n"
+    "  <file/str>\n"
+    "\n"
+    "       File or string containing a CHIP CMS Signed CD message.\n"
+    "\n"
+);
+
+OptionSet *gCmdOptionSets[] =
+{
+    &gCmdOptions,
+    &gHelpOptions,
+    nullptr
+};
+// clang-format on
+
+const char * gInFileNameOrStr = nullptr;
+const char * gOutFileName     = "-";
+
+bool HandleOption(const char * progName, OptionSet * optSet, int id, const char * name, const char * arg)
+{
+    switch (id)
+    {
+    case 'o':
+        gOutFileName = arg;
+        break;
+    default:
+        PrintArgError("%s: Unhandled option: %s\n", progName, name);
+        return false;
+    }
+
+    return true;
+}
+
+bool HandleNonOptionArgs(const char * progName, int argc, char * const argv[])
+{
+    if (argc == 0)
+    {
+        PrintArgError("%s: Please specify the CD to be printed.\n", progName);
+        return false;
+    }
+
+    if (argc > 1)
+    {
+        PrintArgError("%s: Unexpected argument: %s\n", progName, argv[1]);
+        return false;
+    }
+
+    gInFileNameOrStr = argv[0];
+
+    return true;
+}
+
+enum CDFormat
+{
+    kCDFormat_Unknown = 0,
+    kCDFormat_Raw,
+    kCDFormat_Hex,
+    kCDFormat_Base64,
+};
+
+CDFormat DetectCDFormat(const uint8_t * cd, uint32_t cdLen)
+{
+    static const uint8_t cdRawPrefix1[] = { 0x30, 0x81 };
+    static const uint8_t cdRawPrefix2[] = { 0x30, 0x82 };
+    static const char * cdHexPrefix     = "308";
+    static const char * cdB64Prefix     = "MI";
+
+    VerifyOrReturnError(cd != nullptr, kCDFormat_Unknown);
+
+    if ((cdLen > sizeof(cdRawPrefix1)) &&
+        (memcmp(cd, cdRawPrefix1, sizeof(cdRawPrefix1)) == 0 || memcmp(cd, cdRawPrefix2, sizeof(cdRawPrefix2)) == 0))
+    {
+        return kCDFormat_Raw;
+    }
+
+    if ((cdLen > strlen(cdHexPrefix)) && (memcmp(cd, cdHexPrefix, strlen(cdHexPrefix)) == 0))
+    {
+        return kCDFormat_Hex;
+    }
+
+    if ((cdLen > strlen(cdB64Prefix)) && (memcmp(cd, cdB64Prefix, strlen(cdB64Prefix)) == 0))
+    {
+        return kCDFormat_Base64;
+    }
+
+    return kCDFormat_Unknown;
+}
+
+bool ReadCD(const char * fileNameOrStr, MutableByteSpan cd)
+{
+    CDFormat cdFmt = kCDFormat_Unknown;
+    uint32_t cdLen = 0;
+    std::unique_ptr<uint8_t[]> cdBuf;
+
+    // If fileNameOrStr is a file name
+    if (access(fileNameOrStr, R_OK) == 0)
+    {
+        VerifyOrReturnError(ReadFileIntoMem(fileNameOrStr, nullptr, cdLen), false);
+
+        cdBuf = std::unique_ptr<uint8_t[]>(new uint8_t[cdLen]);
+
+        VerifyOrReturnError(ReadFileIntoMem(fileNameOrStr, cdBuf.get(), cdLen), false);
+
+        cdFmt = DetectCDFormat(cdBuf.get(), cdLen);
+        if (cdFmt == kCDFormat_Unknown)
+        {
+            fprintf(stderr, "Unrecognized CD Format in File: %s\n", fileNameOrStr);
+            return false;
+        }
+    }
+    // Otherwise, treat fileNameOrStr as a pointer to the certificate string
```

Comment:
What does "certificate string" mean? Does it mean hex? If so, please say "as hex"

Reply:
updated this comment. The command line input string can be of hex or base64 format